### PR TITLE
Users can comment on stories.

### LINKED ISSUE #5: Users can comment on stories.

### PR DESCRIPTION
**Pull request by connorlay**
resolves #5 #6 #9

Users can comment on stories and reply to other comments. Comments and replies have votes. Users can edit and delete their comments and replies. When a comment is deleted, its replies are also deleted (I may change this to the reddit style of preserving the comment tree).
pr6 1
pr6 2
pr6 3
pr6 4
pr6 5
pr6 6
pr6 7

### CODE REVIEW

File: app/controllers/comments_controller.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  def comment_params
+    params.require(:comment).permit(:commentable_id, :commentable_type, :text)
+  end
+
+  def get_comment
+    @comment = Comment.find_by(id: params[:id])
+    redirect_with_error if @comment.nil?
+  end
+
+  def correct_user?
```
Comment (efatsi) [inline app/controllers/comments_controller.rb]: convention states that methods ending in a `?` return a boolean value. consider naming this something more like `ensure_correct_user`?

File: spec/controllers/comments_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    it "deletes the comment and its votes" do
+      expect {
+        post :destroy, id: comment.id
+      }.to change {
+        Vote.count
+      }.by -5
+      expect(response).to redirect_to("/stories/#{story.id}")
+    end
+    it "deletes the comment's replies" do
```
Comment (efatsi) [inline spec/controllers/comments_controller_spec.rb]: these spec names can be a little more correctly descriptive. the first is just checking if the comment's votes are deleted, the second is testing that the comment and it's replies are being deleted

File: spec/pages/story_page.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+  def has_errors?
+    page.has_content? I18n.t(:errors, count: 1)
+  end
+
+  def path
+    "/stories/#{@story.id}"
+  end
+
+end
```
Comment (efatsi) [inline spec/pages/story_page.rb]: This is really neat. Curious to see if it holds up and is reusable across other specs.

**Comment by efatsi** [comment]
A couple nitpicks, but really nice work overall. :+1:

File: spec/pages/story_page.rb (hunk 3 above)
Comment (connorlay) [inline spec/pages/story_page.rb]: I really hope so! Decoupling Capybara matchers and queries from feature specs helped me a lot with debugging.

File: app/controllers/comments_controller.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  private
+
+  def comment_params
+    params.require(:comment).permit(:commentable_id, :commentable_type, :text)
+  end
+
+  def get_comment
+    @comment = Comment.find_by(id: params[:id])
```
Comment (reagent) [inline app/controllers/comments_controller.rb]: Alternative might be to use `Comment.find` (which will raise `ActiveRecord::RecordNotFound`) and then use [`rescue_from`](http://api.rubyonrails.org/classes/ActiveSupport/Rescuable/ClassMethods.html):

```ruby
rescue_from ActiveRecord::RecordNotFound, :with => :redirect_with_error
```

File: app/controllers/comments_controller.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    @comment = current_user.comments.build(comment_params)
+    if @comment.save
+      redirect_to @comment.story
+    else
+      redirect_with_error
+    end
+  end
+
+  def update
+    if @comment.update_attributes(text: params[:comment][:text])
```
Comment (reagent) [inline app/controllers/comments_controller.rb]: You're checking that this comment belongs to the current user, but you could use a scoped finder instead (assuming `User.has_many :comments`):

```ruby
@comment = current_user.comments.find(params[:id])
```

This will raise `AR::RecordNotFound`, so `rescue_from` will catch this and do the right thing:tm:.

File: app/controllers/comments_controller.rb (hunk 4 above)
Comment (connorlay) [inline app/controllers/comments_controller.rb]: What do developers usually prefer in situations like this? Using methods that raise exceptions or methods that return ```nil```?

Comment (reagent) [inline app/controllers/comments_controller.rb]: I prefer fewer lines unless exceptions are inappropriate (which I think record not found is an exceptional case here)

File: spec/controllers/comments_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        }
+        expect(response).to redirect_to("/stories/#{story.id}")
+      end
+    end
+  end
+
+  describe "DESTROY" do
+
+    it "deletes the comment's votes" do
+      expect {
```
Comment (reagent) [inline spec/controllers/comments_controller_spec.rb]: Hard to parse this with such a small assertion split across multiple lines ... feels about this?

```ruby
expect { post(:destroy, id: comment.id) }.to change { Vote.count }.by(-5)
```

Also, `-5` is a magic number -- where did that come from?

File: spec/controllers/comments_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    request.env["HTTP_REFERER"] = "/stories/#{story.id}"
+    log_in user
+  end
+
+  describe "CREATE" do
+
+    context "with invalid parameters" do
+      let(:params) { { commentable_id: -1, commentable_type: "Story", text: "some text" } }
+
+      it "does not create the comment" do
```
Comment (reagent) [inline spec/controllers/comments_controller_spec.rb]: Is there a spec to see that a comment is created?

Comment (reagent) [inline spec/controllers/comments_controller_spec.rb]: nvm -- didn't realize this was a controller spec.

File: spec/pages/story_page.rb (hunk 3 above)
Comment (reagent) [inline spec/pages/story_page.rb]: Not a fan of this -- looks like a premature refactoring that causes the reader to lose all context within the spec itself.  As a reader of the spec, I would prefer to see local helper methods that might get split out into other files as reuse needs become clearer.

Comment (reagent) [inline spec/pages/story_page.rb]: Though I will say that I like how `expect(story_page).to have_reply(text)` reads ...

Comment (reagent) [inline spec/pages/story_page.rb]: `¯\_(ツ)_/¯`

**Comment by reagent** [comment]
:+1: nice work -- only thing I feel super strong about is the magic numbers in specs (e.g. change by -5). Incorporate or ignore my other comments.

File: spec/controllers/comments_controller_spec.rb (hunk 6 above)
Comment (connorlay) [inline spec/controllers/comments_controller_spec.rb]: ```comment_with_votes``` creates a comment object with 5 votes associated with it. I can explicitly state the ```vote_count``` when I call ```create :comment_with_votes``` to clear up any confusion.

I split up this assertion into multiple lines mostly out of habit... most of the time when I use ```expect { ... }.to change { ... }``` it ends up being too long for me to fit onto one line in good conscience. I figured consistency is better, even if multiple lines are not strictly necessary here?
